[DNM] storage: avoid double copy of RocksDB data

jordanlewis opened the pull request:
Previously, all data returned from RocksDB was allocated for and copied
twice:

1. In C++, from RocksDB's key/value slices to a newly-allocated WriteBatch
2. In Go, from the C++-allocated WriteBatch to a Go-allocated byte slice

This is clearly undesirable, as it's expensive to allocate and copy all
of that data twice.

Now, Go allocates a byte slice based on a rough estimate of the size of
the data and passes that slice into C++, which uses that Go memory
instead of allocating on its own.

If the size of the actual data exceeds the size of the Go byte slice, C++
calls back into Go to request more memory. n.b. this dynamic resizing
happened before this patch as well, localized to C++.

See #21840.

Release note (performance improvement): perform fewer in-memory copies
of result data.

jordanlewis committed:
storage: avoid double copy of RocksDB data

Previously, all data returned from RocksDB was allocated for and copied
twice:

1. In C++, from RocksDB's key/value slices to a newly-allocated WriteBatch
2. In Go, from the C++-allocated WriteBatch to a Go-allocated byte slice

This is clearly undesirable, as it's expensive to allocate and copy all
of that data twice.

Now, Go allocates a byte slice based on a rough estimate of the size of
the data and passes that slice into C++, which uses that Go memory
instead of allocating on its own.

If the size of the actual data exceeds the size of the write batch, C++
calls back into Go to request more memory. n.b. this dynamic resizing
happened before this patch as well, localized to C++.

Release note (performance improvement): perform fewer in-memory copies
of result data.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21855)
<!-- Reviewable:end -->

jordanlewis commented:
This currently has a bug in it somewhere, causing some sporadic test failures.

But assuming this approach can be made workable, we should see some nice performance improvements on large scans. Here's two `COUNT(*)` runs on 100 warehouses, before and after, showing better than a 2x improvement:

```
root@:26257/> select count(*) from tpcc.stock where s_w_id<100;
+----------+
|  count   |
+----------+
| 10000000 |
+----------+
(1 row)

Time: 28.112353833s
```

```
root@:26257/> select count(*) from tpcc.stock where s_w_id<100;
+----------+
|  count   |
+----------+
| 10000000 |
+----------+
(1 row)

Time: 13.649053374s
```

jordanlewis commented:
I think despite the lack of dynamic check failures from CGo there's an illegal operation here which is causing a race with the GC. Namely after growSlice, only the C copy of the input DBSlice has a reference to the grown slice. To prevent this, we would need to pass a pointer to the input DBSlice from Go in the first place, but that's also illegal as CGo prevents passing Go pointers to memory with other Go pointers inside.

Probably a different strategy is necessary. We could hold a synchronized map of grown slices in Go, adding to the set in  growSlice, passing their map indexes back to C, and having the original invoker of C.MVCCScan remove the grown slices from the map once C is finished. Maybe there's a better way as that seems messy.

petermattis commented:
@jordanlewis Right, this is exactly the problem I was referring to in https://github.com/cockroachdb/cockroach/issues/21840#issuecomment-361117217. Using a synchronized map on the Go side is likely going to remove much of the performance benefit. The prohibition on passing a Go struct to C which contains pointers can be hacked around (I think). The other option is to have the C++ MVCCScan call return when it needs a bigger slice, but that will require an irritating amount of refactoring, though clearly the perf benefit looks worth it if we can't find another way.

petermattis committed:
storage/engine: hold a ref to the grown slice

Unfortunately, the results are less than stellar.

name                                                   old time/op    new time/op    delta
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=8-8       253µs ± 2%     229µs ± 0%   -9.32%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-8      269µs ± 2%     280µs ± 1%   +3.96%  (p=0.000 n=9+9)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-8     429µs ± 1%     495µs ± 1%  +15.43%  (p=0.000 n=10+10)

name                                                   old speed      new speed      delta
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=8-8    31.6MB/s ± 2%  34.9MB/s ± 0%  +10.26%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-8    238MB/s ± 2%   229MB/s ± 1%   -3.82%  (p=0.000 n=9+9)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-8  1.19GB/s ± 1%  1.03GB/s ± 1%  -13.36%  (p=0.000 n=10+10)

Release note: None

petermattis commented:
@jordanlewis See https://github.com/petermattis/cockroach/tree/pmattis/mvcc-scan-result-memory. Unfortunately, the `MVCCScan` benchmarks get worse. The size of the initial buffer allocated on the Go side makes a difference. Too large and performance gets worse. Too small and we don't see a performance improvement (or performance gets worse).

jordanlewis commented:
Hmm, I think there's a way to get a good result here still. I see that you got rid of the batch size heuristic for the initial slice size. I'm going to try to add that back in.

jordanlewis committed:
use append instead of make

Release note: None

petermattis commented:
> Hmm, I think there's a way to get a good result here still. I see that you got rid of the batch size heuristic for the initial slice size. I'm going to try to add that back in.

In my experimentation, allocating a slice that is too large initially hurts performance. We'd see this in the `MVCCScan` benchmarks for smaller sizes. Actually, we probably wouldn't see it as the value sizes used in the benchmarks are predictable.

jordanlewis commented:
I could see that, but I think we have a decent enough approximate lower bound which is the size of the batch times the length of the start key. Obviously it's not perfect because key sizes aren't homogeneous, but I think it's worth trying to take into account.

petermattis commented:
> I could see that, but I think we have a decent enough approximate lower bound which is the size of the batch times the length of the start key. Obviously it's not perfect because key sizes aren't homogeneous, but I think it's worth trying to take into account.

Does SQL pass in a good estimate of the number of keys to read? I recall it didn't, but I could be mistaken.

jordanlewis commented:
The KVFetcher has a precise number of keys that it requests per batch. The batch size starts small and increases until it hits 10,000. SQL (via the KV fetcher) never asks for an unbounded number of keys at once.

petermattis commented:
> The KVFetcher has a precise number of keys that it requests per batch.

Right. What I can't remember is whether it sizes that number of keys based on the WHERE clause. For example, for the query `SELECT * FROM test.kv WHERE k = 1` is SQL setting the max keys to 1 or something a bit larger?

jordanlewis commented:
There appears to be some logic for that, but I don't understand it. See `scanNode.limitHint` and the `hardLimit` and `softLimit` values. Maybe @RaduBerinde could answer this question without too much digging?

jordanlewis commented:
I'm also experiencing difficulty trying to tune the initial buffer size.

I'm surprised that this is particularly hard, though. I would have thought that eliminating half of the allocations and copies would have a good effect despite moving some of the allocation to Go. I wonder if the overhead is dominated by the extra CGo crossings or just the added slowness of Go allocation vs C++ allocation.

petermattis commented:
Go has to zero initialize memory. I'm wondering if that is affecting the Go allocation.

jordanlewis committed:
adaptive resizing

Release note: None

jordanlewis commented:
The profile says it takes 10% of the time.

I tried changing `growSlice` to use `append` (with both a zeroed fresh slice and the input slice as the second argument) instead of `make` + `copy`, since `append` doesn't zero the memory it modifies first.

This helped, removing the performance penalty:

```
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=8-8       324µs ± 2%     270µs ± 3%   -16.63%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-8      338µs ± 1%     293µs ± 2%   -13.20%  (p=0.000 n=10+10)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-8     500µs ± 1%     486µs ± 3%    -2.82%  (p=0.003 n=9+10)
```

Then I tried changing `growSlice` to take the current number of seen keys and the maximum number of keys as arguments, so it can estimate the size of the final slice assuming that the keys so far are a representative sample of the rest of the scan (yes, that's not a great assumption). Now the benchmark results are much better:

```
name                                                   old time/op    new time/op    delta
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=8-8       321µs ± 2%     274µs ± 1%   -14.69%  (p=0.000 n=10+9)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-8      339µs ± 1%     293µs ± 1%   -13.40%  (p=0.000 n=9+8)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-8     499µs ± 1%     402µs ± 1%   -19.45%  (p=0.000 n=9+10)

name                                                   old speed      new speed      delta
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=8-8    24.9MB/s ± 2%  29.2MB/s ± 1%   +17.21%  (p=0.000 n=10+9)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=64-8    189MB/s ± 1%   218MB/s ± 1%   +15.48%  (p=0.000 n=9+8)
MVCCScan_RocksDB/rows=1000/versions=1/valueSize=512-8  1.03GB/s ± 1%  1.27GB/s ± 1%   +24.15%  (p=0.000 n=9+10)
```

However I'm nervous that the "adaptive resizing" is a horrible heuristic and will break under real workloads. For example, a very large key right in the beginning of a scan batch would cause a giant slice to get allocated. Perhaps we should just put a cap on the size of new slices created by this resizing.

petermattis commented:
Adaptive resizing was something I thought of too. Glad you implemented it.

Another thought is that we should avoid the copying in `growSlice` completely. Really we're trying to return a chunk of data to Go. We could replace `growSlice` with `newSlice` and have `dataRef` keep references to all of the slices. There would be more work to decode the entries as we'd have N WriteBatches instead of 1, but I think that isn't too bad.

jordanlewis committed:
limit the size of the guess to 16mb more

Release note: None

jordanlewis commented:
Interesting idea about multiple WriteBatches. I would be a little concerned that it would give bad cache behavior later in the system, because the data would no longer be contiguous, but we probably do enough other Bad Things right now that this doesn't matter.

I think in a future where we have a directly castable PAX-ish format, we would definitely want a contiguous data batch. Now I'm not so sure it matters one way or the other, but doing adaptive resizing is easier and with a cap on the max size I'm not sure what the downsides are if it looks decent in profiles.

Regardless, I just checked performance of the KV load generator on this patch and it absolutely tanks the performance (19k ops/s vs 11k ops/s) despite never having to call back into Go to get more memory. So I don't know if this is at all a workable approach.

jordanlewis commented:
Oh but this poor performance is probably related to the limit hint propagation we we're discussing before. I bet if we could teach the scanner to provide a better estimate for the batch size the regression would go away.

nvanbenschoten commented:
Really cool stuff here @jordanlewis!

> Interesting idea about multiple WriteBatches. I would be a little concerned that it would give bad cache behavior later in the system, because the data would no longer be contiguous, but we probably do enough other Bad Things right now that this doesn't matter.

Perhaps we could take a middle ground and immediately combine all `WriteBatches` once the scan is complete and we're back in Go. That would bring us from a logarithmic number of memory copies to a constant number.

petermattis commented:
> Interesting idea about multiple WriteBatches. I would be a little concerned that it would give bad cache behavior later in the system, because the data would no longer be contiguous, but we probably do enough other Bad Things right now that this doesn't matter.

As mentioned offline, cache effects are real but they apply at sizes near the cache line size (64-128 bytes). We're talking about buffers that are 10s to 100s of kilobytes. Have a handful of such buffers should be equivalent for performance vs having one. 

> I think in a future where we have a directly castable PAX-ish format, we would definitely want a contiguous data batch. Now I'm not so sure it matters one way or the other, but doing adaptive resizing is easier and with a cap on the max size I'm not sure what the downsides are if it looks decent in profiles.

A PAX-ish approach doesn't require a single contiguous buffer. I'm not sure if it is mentioned in the PAX paper, but database page sizes are in the 4KB-32KB range. We configure RocksDB with a 32KB block size. That much contiguous data is more than enough to get good cache performance and amortize the overhead of setting up loops.

> Regardless, I just checked performance of the KV load generator on this patch and it absolutely tanks the performance (19k ops/s vs 11k ops/s) despite never having to call back into Go to get more memory. So I don't know if this is at all a workable approach.

Glad you checked that. Performance work is hard, especially in an area of the code that has seen a lot of attention already. 

> Perhaps we could take a middle ground and immediately combine all WriteBatches once the scan is complete and we're back in Go. That would bring us from a logarithmic number of memory copies to a constant number.

That doesn't sound too much different than having C++ allocate memory and do a single copy in Go. Perhaps I'm misunderstanding the suggestion.

nvanbenschoten commented:
> That doesn't sound too much different than having C++ allocate memory and do a single copy in Go. Perhaps I'm misunderstanding the suggestion.

Looking at it that way, the suggestion is actually orthogonal to what we're doing here. The idea is to avoid the copying in `growSlice`, like you brought up earlier, by maintaining multiple buffers in C++. We'd probably want each buffer to be twice the size of the previous one, but we could also try to tune the heuristic to make better guesses like `growSlice` does now. We would then combine them all at the end into a single buffer, where we'd know the exact size necessary to fit all data in the `WriteBatch`. This combined buffer is the only one that would need to be accessible to Go, so it would allow us to reduce the C++ -> Go calls down to just one. 

Assuming `growSlice` did the traditional double-and-copy strategy instead of its current strategy, I believe it currently performs O(log n) memcpys, it copies O(n) bytes in total, and it requires O(log n) C++ -> Go crossings. By maintaining handles on multiple buffers at a time in C++, I think we can reduce this to O(1) memcpys, O(n) bytes copied, and O(1) C++ -> Go crossings.

petermattis commented:
@nvanbenschoten Earlier experimentation with pre-allocating the correct size buffer on the C++ side didn't show a significant performance improvement (small single digits). The big win is to avoid any copying from C++ memory to Go memory by encoding directly into Go memory. Of course, we might have messed something up with the earlier experimentation.

nvanbenschoten commented:
> Earlier experimentation with pre-allocating the correct size buffer on the C++ side didn't show a significant performance improvement (small single digits).

So in this experiment we allocated exactly the correct size buffer in C++ and only did a single copy over to Go? I'm very surprised that performed worse than what we're doing here. Even in this PR's approach, the number of bytes copied is still [asymptotically approaching](https://github.com/cockroachdb/cockroach/pull/21855/files#diff-a0530d32ec5dde00c2e61906032e5f81R127) the number of total bytes in the buffer.

Did we experiment with pre-allocating the correct buffer size on the Go side and encoding directly into that buffer without ever needing to grow it? That would show us the theoretical maximum speedup we can try to achieve here. (sorry if this was already discussed on Slack)

petermattis commented:
> So in this experiment we allocated exactly the correct size buffer in C++ and only did a single copy over to Go? I'm very surprised that performed worse than what we're doing here. Even in this PR's approach, the number of bytes copied is still asymptotically approaching the number of total bytes in the buffer.

Yes, that's correct: we allocated just a bit larger than the correct size buffer in C++ and only performed a single copy over to Go. Agreed that the lack of significant performance improvement is strange. I feel like we're missing something here. Or we made a mistake somewhere. 

> Did we experiment with pre-allocating the correct buffer size on the Go side and encoding directly into that buffer without ever needing to grow it? That would show us the theoretical maximum speedup we can try to achieve here. (sorry if this was already discussed on Slack)

Not quite. We did experiment with allocating the correct buffer size in C++ and not copying it to Go but using it directly. This was only possible due to the known lifetime of that buffer for the benchmarks. See https://github.com/cockroachdb/cockroach/issues/21840#issuecomment-360946417.

jordanlewis commented:
> Oh but this poor performance is probably related to the limit hint propagation we we're discussing before. I bet if we could teach the scanner to provide a better estimate for the batch size the regression would go away.

The KV workload indeed asks for unlimited-size batches. Not great! I will look into whether it's feasible to fix the request batch size for precise gets.
